src/mpi/misc/memory_alloc_kinds.c

```diff
@@ -5,47 +5,60 @@
 
 #include "mpiimpl.h"
 
-static const char *memory_alloc_kinds[6][5] = {
-    /* mpi 4.1 kinds */
-    {"mpi", "alloc_mem", "win_allocate", "win_allocate_shared", NULL},
-    {"system", NULL},
-    /* memory allocation kinds 1.0 */
-    {"cuda", "host", "device", "managed", NULL},
-    {"rocm", "host", "device", "managed", NULL},
-    {"level_zero", "host", "device", "shared", NULL},
-    {NULL}
-};
+/* mpi 4.1 kinds */
+static const char *mpi_kinds[5] = {"mpi", "alloc_mem", "win_allocate", "win_allocate_shared", NULL};
+static const char *system_kinds[2] = {"system", NULL};
+/* memory allocation kinds side doc 1.0 */
+static const char *cuda_kinds[5] = {"cuda", "host", "device", "managed", NULL};
+static const char *rocm_kinds[5] = {"rocm", "host", "device", "managed", NULL};
+static const char *ze_kinds[5] = {"level_zero", "host", "device", "shared", NULL};
+
+/* we support a maximum of 3 kinds at a time: mpi,system,gpu */
+static const char **memory_alloc_kinds[4] = {mpi_kinds, system_kinds, NULL, NULL};
```

Comment:
I was thinking that if you remove the `const`, then you can simply reset unavailable entries to `NULL` at init time.

Reply:
I started down that route, but shifted to an additive model to keep the code for iterating over the supported kinds simpler. I could go either way. Unless we plan on supporting multiple GPU kinds at a time, it probably won't matter.

Reply:
Is this acceptable or should I go back and convert the initialization to negate unsupported types instead?